File: 2020/3.scala
import scala.io.Source

val map = Source.fromFile("3.txt").getLines().toSeq

def trees(slope: (Int, Int)): Long = (map.indices by slope._2).zipWithIndex.count{case (y, i) => map(y)(i*slope._1%map(y).length) == '#'}
println(trees((3, 1)))
println(Seq((1, 1), (3, 1), (5, 1), (7, 1), (1, 2)).map(trees).product)

File: 2020/3.py
import math

with open('3.txt') as f:
    _map = f.read().splitlines()

def trees(slope):
    t = 0
    for i, y in enumerate(range(0, len(_map), slope[1])):
        if _map[y][i*slope[0]%len(_map[y])] == '#':
            t += 1
    return t

print(trees((3, 1)))
print(math.prod(trees(slope) for slope in [(1, 1), (3, 1), (5, 1), (7, 1), (1, 2)]))
